### PR TITLE
Backend Plugins: PluginContext and exported protobuf converters

### PR DESCRIPTION
**Pull request by kylebrandt**
newer SDK, breaking change.

Goes with Marcus' https://github.com/grafana/grafana-plugin-sdk-go/pull/141
and https://github.com/grafana/grafana-plugin-sdk-go/pull/148

### CODE REVIEW

File: pkg/plugins/backendplugin/manager.go [hunk 1]
```diff
@@ -222,11 +235,13 @@ func (m *manager) CallResource(config PluginConfig, reqCtx *models.ReqContext, p
 
 	clonedReq := reqCtx.Req.Clone(reqCtx.Req.Context())
 	keepCookieNames := []string{}
-	if config.JSONData != nil {
-		if keepCookies := config.JSONData.Get("keepCookies"); keepCookies != nil {
-			keepCookieNames = keepCookies.MustStringArray()
-		}
-	}
+	/// TODO! THIS SEEMS IMPORTANT BUT I DO NOT GET IT
```
Comment (kylebrandt) [inline pkg/plugins/backendplugin/manager.go:238]: Don't skip this.

File: pkg/api/plugins.go [hunk 2]
```diff
@@ -21,35 +20,37 @@ import (
...
+	// 	decryptedSecureJSONData = make(map[string]string)
+	// } else {
+	// 	decryptedSecureJSONData = ps.DecryptedValues()
+	// 	updated = ps.Updated
+	// }
+
+	return backend.PluginContext{
+		OrgID:    user.OrgId,
+		PluginID: plugin.Id,
+		// TODO: Which thing?
```
Comment (marefr) [inline pkg/api/plugins.go:50]: `AppInstanceSettings`, this is for `/api/plugins/<plugin id>/<metrics|health|resources>` endpoints

File: pkg/api/plugins.go [hunk 3]
```diff
@@ -356,8 +357,12 @@ func (hs *HTTPServer) CallResource(c *models.ReqContext) {
 		c.JsonApiErr(500, "Failed to get plugin settings", err)
 		return
 	}
-
-	hs.BackendPluginManager.CallResource(config, c, c.Params("*"))
+	// pCtx := &backend.PluginContext{
+	// 	OrgID:    c.OrgId,
+	// 	PluginID: pluginID,
+	// 	// TODO: App instance settings?
```
Comment (marefr) [inline pkg/api/plugins.go:363]: Yes, see earlier comment.

File: pkg/plugins/backendplugin/manager.go (hunk 1 above)
Comment (marefr) [reply pkg/plugins/backendplugin/manager.go:238]: You can configure this in data source config if you're using the HTTPSettings react component in your data source plugin config editor. Should check `DataSourceInstanceSettings` and proceed if not nil.

**Comment by kylebrandt** [comment]
Locally I now get 

```$ go get github.com/grafana/grafana-plugin-sdk-go@7a5404a
can't load package: package github.com/grafana/grafana-plugin-sdk-go: build constraints exclude all Go files in /home/kbrandt/go/pkg/mod/github.com/grafana/grafana-plugin-sdk-go@v0.53.1-0.20200422160825-7a5404ad513e```

not sure why.

**Comment by kylebrandt** [comment]
This did remove 100 lines of code before vendor :-/

File: pkg/api/datasources.go [hunk 4]
```diff
@@ -296,23 +298,17 @@ func (hs *HTTPServer) CallDatasourceResource(c *models.ReqContext) {
...
-		},
-	}
-	hs.BackendPluginManager.CallResource(config, c, c.Params("*"))
+	dsInstanceSettings, err := wrapper.ModelToInstanceSettings(ds)
+	if err != nil {
+		c.JsonApiErr(500, "Unable to process datasource instance model", err)
+	}
+
+	pCtx := backend.PluginContext{
+		OrgID:                      c.OrgId,
```
Comment (marefr) [inline pkg/api/datasources.go:307]: Think user is missing here?

File: pkg/api/datasources.go [hunk 5]
```diff
@@ -367,24 +363,17 @@ func (hs *HTTPServer) CheckDatasourceHealth(c *models.ReqContext) {
...
-		},
-	}
-
-	resp, err := hs.BackendPluginManager.CheckHealth(c.Req.Context(), config)
+	dsInstanceSettings, err := wrapper.ModelToInstanceSettings(ds)
+	if err != nil {
+		c.JsonApiErr(500, "Unable to get datasource model", err)
+	}
+	pCtx := backend.PluginContext{
+		OrgID:                      c.OrgId,
```
Comment (marefr) [inline pkg/api/datasources.go:371]: Think user is missing here?

File: pkg/api/plugins.go [hunk 6]
```diff
@@ -7,49 +7,54 @@ import (
...
 
-	return backendplugin.PluginConfig{
-		OrgID:                   user.OrgId,
-		PluginID:                plugin.Id,
-		JSONData:                jsonData,
-		DecryptedSecureJSONData: decryptedSecureJSONData,
-		Updated:                 updated,
+	return backend.PluginContext{
+		OrgID:    user.OrgId,
+		PluginID: plugin.Id,
```
Comment (marefr) [inline pkg/api/plugins.go:52]: Think user is missing here?

File: pkg/plugins/datasource/wrapper/datasource_plugin_wrapper_v2.go [hunk 7]
```diff
@@ -25,41 +25,42 @@ type DatasourcePluginWrapperV2 struct {
...
-				User:                    ds.User,
-				BasicAuthEnabled:        ds.BasicAuth,
-				BasicAuthUser:           ds.BasicAuthUser,
-				JsonData:                jsonDataBytes,
-				DecryptedSecureJsonData: ds.DecryptedValues(),
-			},
+		PluginContext: &pluginv2.PluginContext{
+			OrgId:                      ds.OrgId,
+			PluginId:                   tw.pluginId,
+			User:                       backend.ToProto().User(backendplugin.BackendUserFromSignedInUser(query.User)),
```
Comment (marefr) [inline pkg/plugins/datasource/wrapper/datasource_plugin_wrapper_v2.go:58]: This looks wrong. Would suggest populating User field in api package

File: vendor/modules.txt [hunk 8]
```diff
@@ -8,13 +8,13 @@ github.com/BurntSushi/toml
 # github.com/VividCortex/mysqlerr v0.0.0-20170204212430-6c6b55f8796f
 ## explicit
 github.com/VividCortex/mysqlerr
-# github.com/apache/arrow/go/arrow v0.0.0-20191025121910-b789226ccb21
+# github.com/apache/arrow/go/arrow v0.0.0-20200403134915-89ce1cadb678
```
Comment (marefr) [inline vendor/modules.txt:11]: Probably something to think about in SDK and use a proper tagged version of arrow?

**Review by marefr — CHANGES_REQUESTED** [review]
Some things need to be sorted out regarding user field. Added some other comments. But should be minor work left.

File: pkg/plugins/datasource/wrapper/datasource_plugin_wrapper_v2.go (hunk 7 above)
Comment (kylebrandt) [reply pkg/plugins/datasource/wrapper/datasource_plugin_wrapper_v2.go:58]: @marefr I think the Query call and the wrapper I need to keep the signature? If that is so I need to grab it from the tsdbQuery.

File: vendor/modules.txt (hunk 8 above)
Comment (kylebrandt) [reply vendor/modules.txt:11]: Apparently, as of 1 hour ago https://twitter.com/ApacheArrow/status/1253347469313953793

File: pkg/plugins/transform_plugin.go [hunk 9]
```diff
@@ -77,7 +77,9 @@ type TransformWrapper struct {
 
 func (tw *TransformWrapper) Transform(ctx context.Context, query *tsdb.TsdbQuery) (*tsdb.Response, error) {
 	pbQuery := &pluginv2.QueryDataRequest{
-		Config:  &pluginv2.PluginConfig{},
+		PluginContext: &pluginv2.PluginContext{
+			// TODO: Things probably
```
Comment (marefr) [inline pkg/plugins/transform_plugin.go:81]: This looks interesting :smiley: No blocker for merging this, but lets discuss tomorrow

**Review by marefr — APPROVED** [review]
LGTM

Maybe some minor things I would like to change, but no blocker. Lets get this merged

File: pkg/plugins/transform_plugin.go (hunk 9 above)
Comment (kylebrandt) [reply pkg/plugins/transform_plugin.go:81]: Aye, will get to looking at this probably when I look into updating gel to work with this.

**Pull request merged**